test/unittests/execute_call_depth_test.cpp

```diff
@@ -0,0 +1,327 @@
+    EXPECT_THAT(execute(*executor, 1, {}, DepthLimit - 1), Traps());
+    EXPECT_THAT(execute(*executor, 1, {}, DepthLimit), Traps());
+}
+
+TEST(execute_call_depth, call_imported_host_function)
+{
+    /* wat2wasm
+    (func $host_f (import "host" "f") (result i32))
+    (func (result i32) (call $host_f))
+    */
+    const auto wasm =
+        from_hex("0061736d010000000105016000017f020a0104686f737401660000030201000a0601040010000b");
+
+    static int recorded_depth;
+    constexpr auto host_f = [](std::any&, Instance&, const Value*, int depth) noexcept {
+        recorded_depth = depth;
+        return ExecutionResult{Value{1}};
+    };
+
+    const auto module = parse(wasm);
+    auto instance = instantiate(*module, {{{host_f}, module->typesec[0]}});
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}), Result(1_u32));
+    EXPECT_EQ(recorded_depth, 1);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit - 2), Result(1_u32));
+    EXPECT_EQ(recorded_depth, DepthLimit - 1);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit - 1), Traps());
+    EXPECT_EQ(recorded_depth, -1000);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit), Traps());
+    EXPECT_EQ(recorded_depth, -1000);
+}
+
+TEST(execute_call_depth, call_host_function_calling_wasm_function)
+{
+    // Test for "wasm-host-wasm" sandwich.
+    // The host function is obligated to bump depth and pass it along.
+
+    /* wat2wasm
+    (func $host_f (import "host" "f") (result i32))
+    (func (result i32) (call $host_f))
+    (func $leaf (result i32) (i32.const 1))
+    */
+    const auto wasm = from_hex(
+        "0061736d010000000105016000017f020a0104686f73740166000003030200000a0b02040010000b040041010"
+        "b");
+
+    static int recorded_depth;
+    constexpr auto host_f = [](std::any&, Instance& instance, const Value*, int depth) noexcept {
+        recorded_depth = depth;
+        return fizzy::execute(instance, 2, {}, depth + 1);
+    };
+
+    const auto module = parse(wasm);
+    auto instance = instantiate(*module, {{{host_f}, module->typesec[0]}});
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}), Result(1_u32));
+    EXPECT_EQ(recorded_depth, 1);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit - 3), Result(1_u32));
+    EXPECT_EQ(recorded_depth, DepthLimit - 2);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit - 2), Traps());
+    EXPECT_EQ(recorded_depth, DepthLimit - 1);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit - 1), Traps());
+    EXPECT_EQ(recorded_depth, -1000);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit), Traps());
+    EXPECT_EQ(recorded_depth, -1000);
+}
+
+
+/// Infinite recursion
+
+TEST(execute_call_depth, execute_internal_function_infinite_recursion)
```

Comment:
Hmm, can you add a case similar to this where recursion starts at the start function? Generally also cases where start function is present and calls other things.

Reply:
Added.